contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,473 @@
+  * The Flex volume plugin exists as an exec based mechanism to create “out-of-tree” volume plugins.
+  * Because Flex drivers exist and depend on the Flex interface, it will continue to be supported with a stable API.
+  * The CSI Volume plugin will co-exist with Flex volume plugin.
+
+## Design Overview
+
+To support CSI Compliant Volume plugins, a new in-tree CSI Volume plugin will be introduced in Kubernetes. This new volume plugin will be the mechanism by which Kubernetes users (application developers and cluster admins) interact with external CSI volume drivers.
+
+The `SetUp`/`TearDown` calls for the new in-tree CSI volume plugin will directly invoke `NodePublishVolume` and `NodeUnpublishVolume` CSI RPCs through a unix domain socket on the node machine.
+
+Provision/delete and attach/detach must be handled by some external component that monitors the Kubernetes API on behalf of a CSI volume driver and invokes the appropriate CSI RPCs against it.
+
+To simplify integration, the Kubernetes team will offer a containers that captures all the Kubernetes specific logic and act as adapters between third-party containerized CSI volume drivers and Kubernetes (each deployment of a CSI driver would have it’s own instance of the adapter).
+
+## Design Details
+
+### Third-Party CSI Volume Drivers
+
+Kubernetes is as minimally prescriptive on the packaging and deployment of a CSI Volume Driver as possible. Use of the *Communication Channels* (documented below) is the only requirement for enabling an arbitrary external CSI compatible storage driver in Kubernetes.
+
+This document recommends a standard mechanism for deploying an arbitrary containerized CSI driver on Kubernetes. This can be used by a Storage Provider to simplify deployment of containerized CSI compatible volume drivers on Kubernetes (see the “Recommended Mechanism for Deploying CSI Drivers on Kubernetes” section below). This mechanism, however, is strictly optional.
+
+### Communication Channels
+
+#### Kubelet to CSI Driver Communication
+
+Kubelet (responsible for mount and unmount) will communicate with an external “CSI volume driver” running on the same host machine (whether containerized or not) via a Unix Domain Socket.
+
+CSI volume drivers should create a socket at the following path on the node machine: `/var/lib/kubelet/plugins/csi/sockets/[driverName]/kubeletproxy.sock`. For alpha, kubelet will assume this is the location for the Unix Domain Socket to talk to the CSI volume driver. For the beta implementation, we can consider using the [Device Plugin Unix Domain Socket Registration](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md#unix-socket) mechanism to register the Unix Domain Socket with kubelet. This mechanism would need to be extended to support registration of both CSI volume drivers and device plugins independently.
```

Comment:
directories under plugins should be fully qualified, so either `kubernetes.io/csi` or `kubernetes.io~csi` (I forget if we got the escaping right).  Or even `.../container-storage-interface`.

Reply:
This doesn't leave a lot of room for driver daemons that implement multiple interfaces, since it is rooted under "csi".  E.g. imagine a driver that was both storage and device.

Reply:
Why does the socket end up named "kubeletproxy" ?  That just seems weird.  What is being proxied?

Reply:
Good points, I am going to change the path to: ``/var/lib/kubelet/plugins/[SanitizedCSIDriverName]/csi.sock`.

For the alpha implementation, these CSI paths will be hard coded like this. For the beta implementation, we will enable the driver to register the socket with kubelet -- this will allow the driver to have a custom socket path and use the same socket for both device and storage.